src/robot/running/keywordrunner.py

```diff
@@ -206,16 +249,69 @@ def _to_number_with_arithmetics(self, item):
             raise TypeError("Expected number, got %s." % type_name(item))
         return number
 
-    def _run_one_round(self, data, values):
-        name = ', '.join(format_assign_message(var, item)
-                         for var, item in zip(data.variables, values))
-        result = KeywordResult(kwname=name,
-                               type=data.FOR_ITEM_TYPE)
-        for var, value in zip(data.variables, values):
-            self._context.variables[var] = value
-        runner = KeywordRunner(self._context, self._templated)
-        with StatusReporter(self._context, result):
-            runner.run_keywords(data.keywords)
+
+class ForInZipLoopRunner(BasicForLoopRunner):
+    def __init__(self, forstep, templated=False):
+        super(ForInZipLoopRunner, self).__init__(forstep, templated)
+
+    def _flavor_name(self):
+        return 'IN ZIP'
+
+    def _transform_items(self, items):
+        answer = list()
+        for item in items:
+            if not is_list_like(item):
+                # Checking here because _validate() doesn't have access
+                # to the actual values.
+                raise DataError("For-In-Zip Loop items must all be List-like (but not Strings); got %s with value '%s'" % (item.__class__, item))
+        for zipped_item in zip(*items):
+            answer.extend(zipped_item)
+        return answer
+
```

Comment:
@pekkaklarck for stuff like this maybe you could set up a static analysis and/or style checker to run in CI along with the tests. Not sure if it would be worth the effort for you though.